alternator: fill UnprocessedKeys for failed batch reads

psarna opened the pull request:
DynamoDB protocol specifies that when getting items in a batch
failed only partially, unprocessed keys can be returned so that
the user can perform a retry.
Alternator used to fail the whole request if any of the reads failed,
but right now it instead produces the list of unprocessed keys
and returns them to the user, as long as at least 1 read was
successful.

This series comes with a test based on Scylla's error injection mechanism, and thus is only useful in modes which come with error injection compiled in. In release mode, expect to see the following message:
SKIPPED (Error injection not enabled in Scylla - try compiling in dev/debug/sanitize mode)

Fixes #9984

nyh commented on alternator/executor.cc:
```diff
@@ -3145,19 +3145,58 @@ future<executor::request_return_type> executor::batch_get_item(client_state& cli
     // try again. Note that simply a missing key is *not* an error (we already
```
Github doesn't let me put a comment on earlier lines, but note that a few lines up this comment is a FIXME about UnprocessedKeys, so now that you did this, you should remove the FIXME (but not the entire comment).

nyh replied:
I think you missed this comment

nyh commented on alternator/executor.cc:
```diff
@@ -3145,19 +3145,58 @@ future<executor::request_return_type> executor::batch_get_item(client_state& cli
+        // If any of the responses failed, the key iterator will be used to insert
+        // an entry into the UnprocessedKeys object, which will ultimately be returned
+        // to the user in case of partial success of BatchGetItem operation.
+        rjson::value::MemberIterator item_it = request_items.MemberBegin();
```
I found the naming "item_it". This "item" in RequestItems is actually a table and all the requests for this table.  So maybe instead of item_it it should be called something that has to do with tables.

psarna replied:
Ok, I agree that it's vague, I'll try to pick a better name

nyh commented on alternator/executor.cc:
```diff
@@ -3145,19 +3145,58 @@ future<executor::request_return_type> executor::batch_get_item(client_state& cli
+                    rjson::add_with_string_name(response["UnprocessedKeys"], rjson::to_string_view(item_it->name), rjson::empty_object());
+                    rjson::value& unprocessed_item = response["UnprocessedKeys"][item_it->name];
+                    for (auto it = item_it->value.MemberBegin(); it != item_it->value.MemberEnd(); ++it) {
+                        if (it->name != "Keys") {
```
If you want just the "Keys" member why use for() instead of operator[] or find() or something like that?

nyh commented on alternator/executor.cc:
```diff
@@ -3145,19 +3145,58 @@ future<executor::request_return_type> executor::batch_get_item(client_state& cli
+                    rjson::value& unprocessed_item = response["UnprocessedKeys"][item_it->name];
+                    for (auto it = item_it->value.MemberBegin(); it != item_it->value.MemberEnd(); ++it) {
+                        if (it->name != "Keys") {
+                            rjson::add_with_string_name(unprocessed_item, rjson::to_string_view(it->name), rjson::copy(it->value));
```
I'm afraid I don't understand this. If we get a failure for any one of the requests from this table, we copy the entire Keys array - possibly many different requested keys - to UnprocessedKeys? Why? Shouldn't it be set here to an empty_array(), and below you copy to it the one failed key? (same as we do below for the sucessful case!)

psarna replied:
The exact opposite! :) Note that the comparison here is `!=`. The format of `UnprocessedKeys` is super verbose, probably because it's expected to be copied later by the user and reused as input to BatchGetItem. In order to adhere with what DynamoDB returns, I need to copy the entire BatchGetItem metadata **except** keys (projection expression, etc.). Then, I add actual unprocessed keys to the structure one by one - creation of the new array happens in line 3175 and the addition in 3177.

nyh replied:
Oh, I missed that...

nyh commented on alternator/executor.cc:
```diff
@@ -3145,19 +3145,58 @@ future<executor::request_return_type> executor::batch_get_item(client_state& cli
+                }
             }
         }
+        elogger.debug("Unprocessed keys: {}", response["UnprocessedKeys"]);
```
do you want to keep this in the code?

psarna replied:
Not sure. It's definitely useful for testing the code manually, but it might be a little too spammy for debug. On the other hand, `trace` log level is where potentially spammy messages belong, so maybe I'll just degrade it to `trace`.

nyh reviewed (commented):
Looks mostly good, but I think maybe (?) I found one bug in the implementation.

I agree with what you said about testing. One thing I would have loved to have - but we don't have now - is a way for testing code to mark specific keys as "broken", so reads to them will fail. Then you'll be able to write tests which check all the different cases:
1. Read only of broken keys
2. Read of broken and normal keys
3. Read of normal keys
4. Read from multiple tables, only in some of them there are broken keys.

But since we don't anything like this, it will probably be an unreasonable effort to start adding it just for the purpose of this test...

denesb commented:
We do have an error injection framework, why can't you use that to write an automatic unit test?

psarna commented:
> We do have an error injection framework, why can't you use that to write an automatic unit test?

Hm, sounds promising, thanks. I'll try to craft a scylla-only test that relies on injection

nyh commented:
> > We do have an error injection framework, why can't you use that to write an automatic unit test?
> 
> Hm, sounds promising, thanks. I'll try to craft a scylla-only test that relies on injection

Playing the devils advocate for a moment, I think it's actually important to write a test that does work on DynamoDB, because that's the only way to verify that our understanding of the DynamoDB documentation is actually correct. One way to write such a test can be to take advantage of the fact that in DynamoDB a single BatchGetItem operation is limited to retrieving up to 100 items or a total of 16 MB of data,  whichever is smaller - and if you read less than 100 items that total *more* than 16MB, you should get only partial results - and UnprocessedKeys. This approach will allow you to test UnprocessedKeys on DynamoDB as well. However, the snag is that Scylla doesn't currently implement this 16MB limit (this is #5944), so unless you want to do #5944 first (and I doubt you do), what you can do is to write now an xfailing test which will work on DynamoDB (and allow you to look at the exact structure of the returned UnprocessedKeys - not just documentation). You can make it pass in Scylla in a manual test by hacking such a limit unofficially just to run this test once.

psarna commented:
@nyh I went with a simpler approach. Your `test_batch_get_item_large` test case already has the logic for parsing and reusing UnprocessedKeys, so I'll just add a single line that *maybe* injects an error in Scylla to force a partial read. "Maybe", because error injection is not enabled in all build modes, but since Jenkins runs tests on all modes, we're covered.

psarna commented:
v2:
 - renamed item_it to table_items_it, so that it's more self-descriptive
 - extended a batch test case with error injection to force partial batch reads, if enabled

nyh commented on test/alternator/util.py:
```diff
@@ -206,3 +208,17 @@ def client_no_transform(client):
+# is not guaranteed to have any effect.
+def maybe_inject_error(dynamodb, err, one_shot=False):
+    if dynamodb.meta.client._endpoint.host.endswith('.amazonaws.com'):
+        return
```
I'm not thrilled about something not being tested with even knowing that it's not being tested. I prefer what we did in with_tracing() in test_tracing.py - where we call `pytest.skip()` the test if we can't enable that feature.  We can call the function `scylla_inject_error()` to emphasize it is Scylla-only.
However this will mean you'll need a second test (or to avoid code duplication, the same test with a pytest.mark.parameterize) that will  do the injection and that test will be skipped when we can't inject.

psarna replied:
I generally agree, but this case is somewhat specific. DynamoDB will also exercise the code path for UnprocessedKeys, but for a different reason - hitting its batch limits. We don't have those in Scylla, so that's why I slipped in an error injection point.

nyh commented on test/alternator/test_batch.py:
```diff
@@ -313,7 +313,7 @@ def test_batch_unprocessed(test_table_s):
 # request signature, and parses a long request (issue #7213) as well as the
 # resulting long write (issue #8183).
-def test_batch_write_item_large(test_table_sn):
+def test_batch_write_item_large(dynamodb, test_table_sn):
```
This change isn't necessary here.

psarna replied:
oops, indeed

nyh commented on test/alternator/test_batch.py:
```diff
@@ -364,10 +364,12 @@ def test_batch_get_item_large(test_table_sn):
     # UnprocessedKeys), so for reliable success of this test we need a loop:
     responses = []
     to_read = { test_table_sn.name: {'Keys': [{'p': p, 'c': c} for c in range(count)], 'ConsistentRead': True } }
+    maybe_inject_error(dynamodb, "alternator_batch_get_item", one_shot=True)
```
As I noted in the previous patch, instead of a "maybe" you can duplicate this test (either really duplicate it or a part of it, or use `@pytest.mark.parametrize` - we haven't used it in alternator tests yet, but the dtest guys have been using it heavily) - and have the test with inject_error() be automatically skipped when it can't be done.

psarna replied:
I could, but note that it's still a "maybe", because it will depend on how Scylla was compiled. Although in this case it's easy to test that - if error injection is disabled, the list of registered injection points will still be empty after registering such a point via API, and then the test case can be skipped. So maybe I'll indeed duplicate this test after all.

nyh reviewed (commented):
Looks good. I just have some suggestions for you to consider.

psarna commented:
v3:
 - removed the fixed FIXME
 - `maybe_inject_error` is renamed to `scylla_inject_error` and forces the test to be skipped if it's not possible to inject an error
 - removed an unnecessary change in BatchWriteItem test

psarna commented:
Hm, I just marked it ready for review, but I'll add one more test case - that a full failure (as in, no requests succeeded) is indeed reported as a failure and not as a regular response with UnprocessedKeys containing all keys.

psarna commented:
v4:
 - added a test case for full batch read failure

nyh commented on test/alternator/test_batch.py:
```diff
@@ -364,6 +364,7 @@ def test_batch_get_item_large(test_table_sn):
     # UnprocessedKeys), so for reliable success of this test we need a loop:
     responses = []
     to_read = { test_table_sn.name: {'Keys': [{'p': p, 'c': c} for c in range(count)], 'ConsistentRead': True } }
+    scylla_inject_error(dynamodb, "alternator_batch_get_item", one_shot=True)
```
I think you accidentally added a scylla_inject_error to this test too?

nyh commented on test/alternator/test_batch.py:
```diff
@@ -402,4 +402,22 @@ def test_batch_get_item_partial(scylla_only, dynamodb, test_table_sn):
+
+# Test that if the batch read failure is total, i.e. all read requests
+# failed, it's reported as an error and not as a regular response with
+# UnprocessedKeys set to all given keys.
```
This makes me wonder what leads us to believe that this **needs to be** the behavior. I mean, if we do a 3-read batch and two of the reads failed, we get one type of error, but if all 3 failed, we get a different error. Why is this important? Do we have a reason to believe that DynamoDB behaves this way?

nyh commented on test/alternator/test_batch.py:
```diff
@@ -402,4 +402,22 @@ def test_batch_get_item_partial(scylla_only, dynamodb, test_table_sn):
+    responses = []
+    to_read = { test_table_sn.name: {'Keys': [{'p': p, 'c': c} for c in range(count)], 'ConsistentRead': True } }
+    # The error injection is permanent, so it will fire for each batch read.
+    scylla_inject_error(dynamodb, "alternator_batch_get_item", one_shot=False)
```
You probably need to undo this error injection, or all later BatchGetItem tests will fail.
I would add a context manager (like we did for the tracing code) so you can't accidentally set an injection and forget to remove it.

psarna replied:
Ouch, yes, definitely. It wasn't a problem with one_shot injection, but it definitely is in the new case

nyh commented on test/alternator/test_batch.py:
```diff
@@ -402,4 +402,22 @@ def test_batch_get_item_partial(scylla_only, dynamodb, test_table_sn):
+    to_read = { test_table_sn.name: {'Keys': [{'p': p, 'c': c} for c in range(count)], 'ConsistentRead': True } }
+    # The error injection is permanent, so it will fire for each batch read.
+    scylla_inject_error(dynamodb, "alternator_batch_get_item", one_shot=False)
+    with pytest.raises(Exception, match="injection"):
```
What type of exception do we get here? An internal server error? Maybe you can match one of DynamoDB's official error names and not just the word "injection"?
Also why Exception instead of ClientError?

psarna replied:
yeah, it's an internal server error, so I decided that the exact type is not very valuable here. I'll try to narrow the type down

psarna commented:
v5:
 - changed the helper function for error injection to be a context manager, which allows unregistering the failure later
 - removed a few incorrect leftovers, like injecting an error in the original test case

nyh commented on test/alternator/util.py:
```diff
@@ -206,3 +209,22 @@ def client_no_transform(client):
+    url = re.sub(r':[0-9]+(/|$)', ':10000', url)
+    url = re.sub(r'^https:', 'http:', url)
+    response = requests.post(f'{url}/v2/error_injection/injection/{err}?one_shot={one_shot}')
+    response = requests.get(f'{url}/v2/error_injection/injection')
```
nitpick (not important enough to delay this series): can we read just {err} and not all the injections?

psarna replied:
no, we only have POST and DELETE ops for a single injection point. It would be nice indeed to read a single point and all details about it, but it's not here yet

psarna commented on test/alternator/util.py:
```diff
@@ -206,3 +209,22 @@ def client_no_transform(client):
+    print("Enabled error injections:", response.content.decode('utf-8'))
+    if response.content.decode('utf-8') == "[]":
+        pytest.skip("Error injection not enabled in Scylla - try compiling in dev/debug/sanitize mode")
+    yield
```
good to know, thanks

nyh replied:
Yes, I was surprised when I found this out for the first time. pytest fixtures don't behave this way - the function is resumed even if the user's code threw an exception. But contextmanager behaves differently, unfortunately.

psarna commented:
v6:
 - made the new context manager exception-safe
